File: strategy/signal/ma_three_cross.py
from dataclasses import dataclass

from core.models.moving_average import MovingAverageType
from core.models.parameter import CategoricalParameter, Parameter, RandomParameter
from strategy.signal.base import Signal, SignalType


@dataclass(frozen=True)
class MA3CrossSignal(Signal):
    type: SignalType = SignalType.Ma3Cross
    smoothing: Parameter = CategoricalParameter(MovingAverageType)
    short_period: Parameter = RandomParameter(5.0, 50.0, 5.0)
    medium_period: Parameter = RandomParameter(50.0, 100.0, 5.0)
    long_period: Parameter = RandomParameter(100.0, 200.0, 10.0)

File: ta_lib/strategies/signal/src/ma_three_cross.rs
use base::prelude::*;
use core::prelude::*;
use shared::{ma_indicator, MovingAverageType};

pub struct MA3CrossSignal {
    ma: MovingAverageType,
    short_period: usize,
    medium_period: usize,
    long_period: usize,
}

impl MA3CrossSignal {
    pub fn new(
        ma: MovingAverageType,
        short_period: f32,
        medium_period: f32,
        long_period: f32,
    ) -> Self {
        Self {
            ma,
            short_period: short_period as usize,
            medium_period: medium_period as usize,
            long_period: long_period as usize,
        }
    }
}

impl Signal for MA3CrossSignal {
    fn lookback(&self) -> usize {
        let adjusted_lookback = std::cmp::max(self.short_period, self.long_period);
        std::cmp::max(adjusted_lookback, self.medium_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let short_ma = ma_indicator(&self.ma, data, self.short_period);
        let medium_ma = ma_indicator(&self.ma, data, self.medium_period);
        let long_ma = ma_indicator(&self.ma, data, self.long_period);

        (
            short_ma.cross_over(&medium_ma) & medium_ma.sgt(&long_ma),
            short_ma.cross_under(&medium_ma) & medium_ma.slt(&long_ma),
        )
    }
}
